addr_manager_badger.go

```diff
@@ -0,0 +1,282 @@
+	opts := badger.DefaultOptions
+	opts.Dir = dataPath
+	opts.ValueDir = dataPath
+	db, err := badger.Open(opts)
+	if err != nil {
+		return nil, err
+	}
+	return &BadgerAddrManager{DB: db, subsManager: NewAddrSubManager()}, nil
+}
+
+// use murmur3 because it's the most compact
+func hashMultiaddr(addr *ma.Multiaddr) ([]byte, error) {
+	return multihash.Encode((*addr).Bytes(), multihash.MURMUR3)
+}
+
+func createKeyPrefix(p *peer.ID) ([]byte, error) {
+	buf := &bytes.Buffer{}
+	prefix := []byte(*p)
+	prefixlen := uint64(len(prefix))
+	if err := binary.Write(buf, binary.LittleEndian, prefixlen); err != nil {
+		return nil, err
+	}
+	buf.Write(prefix)
+	return buf.Bytes(), nil
+}
+
+func createUniqueKey(p *peer.ID, addr *ma.Multiaddr) ([]byte, error) {
+	buf := &bytes.Buffer{}
+	prefix := []byte(*p)
+	prefixlen := uint64(len(prefix))
+	if err := binary.Write(buf, binary.LittleEndian, prefixlen); err != nil {
+		return nil, err
+	}
+	buf.Write(prefix)
+	addrhash, err := hashMultiaddr(addr)
+	if err != nil {
+		return nil, err
+	}
+	addrhashlen := uint64(len(addrhash))
+
+	if err := binary.Write(buf, binary.LittleEndian, addrhashlen); err != nil {
+		return nil, err
+	}
+
+	buf.Write(addrhash)
+	return buf.Bytes(), nil
+}
+
+func peerIDFromBinaryKey(key []byte) (peer.ID, error) {
+	buf := bytes.NewBuffer(key)
+	idlen, err := binary.ReadUvarint(buf)
+	if err != nil {
+		return peer.ID(""), err
+	}
+	idbytes := make([]byte, idlen)
+	read, err := buf.Read(idbytes)
+	if err != nil {
+		return peer.ID(""), err
+	}
+	if uint64(read) != idlen {
+		return peer.ID(""), errors.New("invalid key length")
+	}
+	return peer.IDFromBytes(idbytes)
+}
+
+func (mgr *BadgerAddrManager) AddAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.AddAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+func (mgr *BadgerAddrManager) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	if ttl <= 0 {
+		log.Debugf("short circuiting AddAddrs with ttl %d", ttl)
+		return
+	}
+
+	mgr.SetAddrs(p, addrs, ttl)
+}
+
+func (mgr *BadgerAddrManager) SetAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.SetAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+func (mgr *BadgerAddrManager) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	txn := mgr.DB.NewTransaction(true)
+	defer txn.Discard()
+
+	for _, addr := range addrs {
+		if addr == nil {
+			continue
+		}
+		key, err := createUniqueKey(&p, &addr)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		if ttl <= 0 {
+			if err := txn.Delete(key); err != nil {
+				log.Error(err)
+				continue
+			}
+		}
+		if item, err := txn.Get(key); err != nil || item.IsDeletedOrExpired() {
+			mgr.subsManager.BroadcastAddr(p, addr)
+		}
+		entry := &addrentry{Addr: addr.Bytes(), TTL: ttl}
+		buf := &bytes.Buffer{}
+		enc := gob.NewEncoder(buf)
+		if err := enc.Encode(entry); err != nil {
+			log.Error(err)
+			continue
+		}
+		txn.SetWithTTL(key, buf.Bytes(), ttl)
+	}
+
+	txn.Commit(nil)
```

Comment:
i'll add a callback for this that logs it in the event of error. when passing a callback, commit returns immediately instead of blocking. i'm not 100% sure what the desired behavior is, but for now i'll make a `done` channel and wait for the callback to execute before returning

Reply:
this actually uncovered a much larger issue re: transaction conflicts. thanks a ton!